---
title: "Single-cell RNA-seq of naive/primed embryonic stem cells: G1-only, primed-only analysis"
author: Tobias Messmer and Aaron Lun
date: 17 December 2016
output: 
  html_document:
    toc: true
    toc_float: true
    depth: 3
    number_sections: true
    theme: united 
    highlight: tango 
    fig_caption: false
---

```{r, echo=FALSE, results="hide"}
dir.create("figure-g1_primed/", showWarning=FALSE)
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE, fig.path="figure-g1_primed/")
options(bitmapType="cairo", width=100)
library(BiocParallel)
register(SerialParam())
```

# Selecting only the primed subpopulations

We select only the primed cells and attempt to identify subpopulations within the primed condition.
This is more sensitive than doing so on the entire data set, where the naive/primed differences dominate.
We subset to only use G1 cells, to check that the cell cycle phase does not have any effect on the analysis.

```{r primedin}
library(scran)
sce <- readRDS("sce_all.rds")
sce <- sce[,sce$phase=="G1"]
sce_primed <- sce[,colData(sce)$phenotype=="primed"] 
dim(sce_primed)
```

We adjust the `batch` factor to get rid of the now-unnecessary "naive" level. 
We also re-normalize to rescale the size factors at unity, so that they are comparable between spike-ins and endogenous genes.

```{r}
sce_primed$batch <- factor(sce_primed$batch)
sce_primed <- normalize(sce_primed)
```

We can examine the expression of the various ESC markers in more detail.

```{r primedExpMarker}
fontsize <- theme(axis.text=element_text(size=12), axis.title=element_text(size=16))
plotExpression(sce_primed, c("KLF4", "KLF17", "DPPA3", "TFCP2L1", "NANOG", "ZFP42", "POU5F1"),
    colour_by="batch") + fontsize
```

We also set up an output directory for the results.

```{r primedout}
resdir <- "results-g1_primed"
dir.create(resdir, showWarning=FALSE)
```

# Detecting primed-only HVGs

We detect HVGs in the primed population, using the variance of the normalized log-expression values.
A trend is fitted to the spike-in variances and represents the technical aspect of variability.
This is done separately for each batch due to the differences in spike-in quantity.
The trend is okay, though some deviation of the spike-ins is attributable to amplification-induced variability.

```{r primed.vartrend, fig.height=6, fig.width=10}
par(mfrow=c(1,2))
collected.var_primed <- list()
for (b in levels(sce_primed$batch)) {
    cur.sce <- sce_primed[,b==sce_primed$batch]
    cur.sce <- normalize(cur.sce)
    var.fit <- trendVar(cur.sce, method="loess", span=0.4)
    var.out <- decomposeVar(cur.sce, var.fit)
    collected.var_primed[[b]] <- var.out 
    plot(var.out$mean, var.out$total, pch=16, cex=0.6, xlab="Mean log-expression",
        ylab="Variance of log-expression")
    points(var.fit$mean, var.fit$var, pch=16, col="red")
    curve(var.fit$trend(x), col="red", add=TRUE, lwd=2)
}
```

We combine statistics across batches (requiring a gene to be HVG in both batches in order to be detected) and save them to file.
We also look at the distribution of the gene-wise variances.
The biological component is computed as the difference in the total variance and the trend (i.e., the technical component).

```{r primed.varplot}
var.out_primed <- do.call(combineVar, collected.var_primed)
write.table(file=file.path(resdir, "var.tsv"), var.out_primed, sep="\t", quote=FALSE, col.names=NA) 
plot(var.out_primed$mean, var.out_primed$total, pch=16, cex=0.6, xlab="Mean log-expression",
     ylab="Variance of log-expression")
is <- which(isSpike(sce_primed))
points(var.out_primed$mean[is], var.out_primed$total[is], col="red", pch=16)
o <- is[order(var.out_primed$mean[is])]
curve(var.fit$trend(x), col="red", add=TRUE, lwd=2)
```

Genes with significantly non-zero biological components at a FDR of 5% are defined as HVGs.

```{r var.primed}
hvg.out_primed <- var.out_primed[which(var.out_primed$FDR <= 0.05 & var.out_primed$bio >= 0.5),]
hvg.out_primed <- hvg.out_primed[order(hvg.out_primed$bio, decreasing=TRUE),]
nrow(hvg.out_primed)
write.table(file=file.path(resdir, "hvg.tsv"), hvg.out_primed, sep="\t", quote=FALSE, col.names = NA) 
head(hvg.out_primed)
```

We check that the HVGs are not being driven by outliers or batch effects.

```{r violin.hvg}
plotExpression(sce_primed, rownames(hvg.out_primed)[1:10], alpha=0.05, 
    colour_by="batch", jitter="jitter") + fontsize
```

# Detecting correlated HVGs

We also assess the pairwise correlations between the genes across the primed cells.
Again, the usual suspects dominate the top set of correlations.

```{r, cor.primed}
set.seed(100) 
design_primed <- model.matrix(~ 0 + sce_primed$batch)
var.cor_primed <- correlatePairs(sce_primed, design = design_primed, 
    subset.row=rownames(hvg.out_primed), per.gene=TRUE)
head(var.cor_primed)
```

We focus on the genes that have significantly large correlations.
This happens to be the majority of HVGs that were originally detected.

```{r, sig.primed}
sig.cor_primed <- var.cor_primed$FDR <= 0.05
summary(sig.cor_primed)
chosen_primed <- var.cor_primed$gene[sig.cor_primed]
saveRDS(file=file.path(resdir, "cor_hvg.rds"), chosen_primed)
length(chosen_primed)
```

```{r, echo=FALSE, results="hide"}
gc()
```

# Performing dimensionality reduction

We remove the batch effect prior to performing dimensionality reduction.

```{r}
library(limma)
norm_exprs(sce_primed) <- removeBatchEffect(exprs(sce_primed), batch=sce_primed$batch) 
```

We then create _t_-SNE plots, using only the set of correlated HVGs.

```{r tsne.primed, fig.width=15, fig.height=6}
tsne1_primed <- plotTSNE(sce_primed, exprs_values="norm_exprs", colour_by=rownames(hvg.out_primed)[1],
    perplexity=5, rand_seed=100, feature_set=chosen_primed) + fontsize
tsne2_primed <- plotTSNE(sce_primed, exprs_values="norm_exprs", colour_by="KLF4",
    perplexity=5, rand_seed=100, feature_set=chosen_primed) + fontsize
tsne3_primed <- plotTSNE(sce_primed, exprs_values="norm_exprs", colour_by="batch",
    perplexity=5, rand_seed=100, feature_set=chosen_primed) + fontsize
multiplot(tsne1_primed, tsne2_primed, tsne3_primed, cols=3)
```

And again, for PCA plots.

```{r pca.primed, fig.width=15, fig.height=6}
pca1_primed <- plotPCA(sce_primed, exprs_values="norm_exprs", colour_by=rownames(hvg.out_primed)[1], 
    feature_set = chosen_primed) + fontsize
pca2_primed <- plotPCA(sce_primed, exprs_values="norm_exprs", colour_by="KLF4", 
    feature_set = chosen_primed) + fontsize
pca3_primed <- plotPCA(sce_primed, exprs_values="norm_exprs", colour_by="batch", 
    feature_set = chosen_primed) + fontsize
multiplot(pca1_primed, pca2_primed, pca3_primed, cols=3)
```

Both plots suggest that there is a small subpopulation off the main bulk of cells.

# Clustering on Euclidean distance

We perform hierarchical clustering on the Euclidean distance, using the correlated HVGs identified previously and the batch-corrected expression values.

```{r primedclust}
chosen.exprs_primed <- norm_exprs(sce_primed)[chosen_primed,]
my.dist_primed <- dist(t(chosen.exprs_primed))
my.tree_primed <- hclust(my.dist_primed, method="ward.D2")
```

To identify the "best" number of clusters, we vary the `k` for tree cutting and examine the resulting silhouette plots.
There's not much evidence for clear separation of the clusters at any `k`.
Many clusters have cells have negative widths that should be allocated to other clusters, indicating that the separation of cells here is quite weak.
As such, we do not perform any clustering.

```{r silhouette, fig.width=10, fig.height=10}
library(cluster)
par(mfrow=c(2,2))
for (k in 2:5){
    my.clusters_primed <- unname(cutree(my.tree_primed, k=k))
    col <- rainbow(k)
    sil <- silhouette(my.clusters_primed, dist = my.dist_primed)
    sil.cols <- col[ifelse(sil[,3] > 0, sil[,1], sil[,2])]
    sil.cols <- sil.cols[order(-sil[,1], sil[,3])]
    plot(sil, main = paste(k, "clusters"), border=sil.cols, 
        col=sil.cols, do.col.sort=FALSE)
}
```

# Comparison of subpopulation heterogeneity

## Between naive and primed conditions

We compare the variability of the naive and primed populations to each other.
We first look at the number of HVGs detected in each population.
The number is slightly lower in the primed population, though that's probably due to loss of power from increases in the technical variance.

```{r}
var.out_naive <- read.table(file.path("results-naive", "var.tsv"))
hvg.out_naive <- read.table(file.path("results-naive", "hvg.tsv"))
nrow(hvg.out_primed)
nrow(hvg.out_naive)
plot(var.out_primed$mean, var.out_primed$tech, col="dodgerblue", pch=16, cex=0.5,
    xlab="Mean log-expression", ylab="Technical variance")
points(var.out_naive$mean, var.out_naive$tech, col="orange", pch=16, cex=0.5)
```

A simpler view is that of the differences in the biological component between conditions.
This is not affected by the technical variance, at least with respect to loss of power.
We plot the difference against the change in the mean expression to see whether DE is driving the change in variance.
(Such genes tend to be difficult to interpret as the biological effect of the change in variability cannot be disentangled from that of differential expression.)

```{r varcomp, fig.width=15, fig.height=6}
dbio <- var.out_primed$bio - var.out_naive$bio

# Expression values are not directly comparable between separately-normalized objects.
# Pulling out primed-naive log-fold changes from the overall analysis instead.
de.stats <- read.table(file.path("results-overall", "de.tsv"))
lfc <- -de.stats[rownames(var.out_naive),"logFC"]

par(mfrow=c(1,3))
phvg <- match(rownames(hvg.out_primed), rownames(var.out_primed))
nhvg <- match(rownames(hvg.out_naive), rownames(var.out_naive))
both.hvg <- intersect(phvg, nhvg)
smoothScatter(lfc[both.hvg], dbio[both.hvg], xlab=expression(Log[2]~"fold change"), 
    ylab=expression(Delta*"biological component"), main="Both HVGs")
phvg.only <- setdiff(phvg, nhvg)
smoothScatter(lfc[phvg.only], dbio[phvg.only], xlab=expression(Log[2]~"fold change"), 
    ylab=expression(Delta*"biological component"), main="Primed-only HVGs")
nhvg.only <- setdiff(nhvg, phvg)
smoothScatter(lfc[nhvg.only], dbio[nhvg.only], xlab=expression(Log[2]~"fold change"), 
    ylab=expression(Delta*"biological component"), main="Naive-only HVGs")
```

We also check the number of expressed genes between naive and primed cells.
The primed cells have more-or-less the same number of expressed genes, which means that the number of genes isn't driving the differences.
(This plot isn't particularly informative because differences in numbers of expressed genes would manifest as a non-zero log-fold change anyway.)

```{r}
naive.genenum <- sum(var.out_naive$mean > 0)
primed.genenum <- sum(var.out_primed$mean > 0)
barplot(cbind(Naive=naive.genenum, Primed=primed.genenum), ylab="Number of expressed genes")
```

```{r, echo=FALSE, results="hide"}
gc()
```

## Interpreting the increased heterogeneity

### Gene set analyses

We assemble genes in every GO term.

```{r gosetup}
library(org.Hs.eg.db)
EG.GO <- toTable(org.Hs.egGO2ALLEGS)
EG.GO <- EG.GO[EG.GO$Ontology=="BP",]
all.sets <- split(EG.GO$gene_id, EG.GO$go_id)
```

We run a one-sided sign-test to identify genes that are more variable in one condition compared to the other.
For example, the p-value for `poutput` represents the probability of obtaining fewer genes that are less variable in the primed condition (i.e., more variability in primed cells).

```{r runsigntest}
poutput <- list()
noutput <- list()
numgenes <- list()
for (entry in names(all.sets)) {
    m <- which(rowData(sce_primed)$entrezgene %in% all.sets[[entry]])
    pbio <- var.out_primed$bio[m]
    nbio <- var.out_naive$bio[m]
    keep <- (nbio > 0 | pbio > 0) & nbio!=pbio
    pbio <- pbio[keep]
    nbio <- nbio[keep]
    numgenes[[entry]] <- sum(keep)
    poutput[[entry]] <- pbinom(sum(pbio < nbio), p=0.5, sum(keep))
    noutput[[entry]] <- pbinom(sum(nbio < pbio), p=0.5, sum(keep))
}
numgenes <- unlist(numgenes)
poutput <- unlist(poutput)
noutput <- unlist(noutput)
```

We summarize the GO terms and save the results to file.

```{r gosumup}
library(GO.db)
go.names <- select(GO.db, keys = names(all.sets), columns = "TERM")
p.res <- data.frame(Term=names(poutput), Description=go.names$TERM, 
    N=numgenes, P=poutput, FDR=p.adjust(poutput, method="BH"))
p.res <- p.res[order(p.res$P),]
write.table(file=file.path(resdir, "go_primed.tsv"), p.res, 
    row.names=FALSE, sep="\t", quote=FALSE)
head(p.res)
n.res <- data.frame(Term=names(noutput), Description=go.names$TERM, 
    N=numgenes, P=noutput, FDR=p.adjust(noutput, method="BH"))
n.res <- n.res[order(n.res$P),]
write.table(file=file.path(resdir, "go_naive.tsv"), n.res, 
    row.names=FALSE, sep="\t", quote=FALSE)
head(n.res)
```

Doing the same for the KEGG terms.

```{r}
EG.KEGG <- toTable(org.Hs.egPATH2EG)
all.sets <- split(EG.KEGG$gene_id, EG.KEGG$path_id)
```

Running the sign tests again.

```{r, ref.label="runsigntest"}
```

Summarizing the results.

```{r keggsumup}
kegg.names <- getKEGGPathwayNames()
kegg.names <- kegg.names[match(paste0("path:map", names(all.sets)), kegg.names$PathwayID),]
p.res <- data.frame(Path=names(poutput), Description=kegg.names$Description, 
     N=numgenes, P=poutput, FDR=p.adjust(poutput, method="BH"))
p.res <- p.res[order(p.res$P),]
write.table(file=file.path(resdir, "kegg_primed.tsv"), p.res, 
    row.names=FALSE, sep="\t", quote=FALSE)
head(p.res)
n.res <- data.frame(Path=names(noutput), Description=kegg.names$Description, 
     N=numgenes, P=noutput, FDR=p.adjust(noutput, method="BH"))
n.res <- n.res[order(n.res$P),]
write.table(file=file.path(resdir, "kegg_naive.tsv"), n.res, 
    row.names=FALSE, sep="\t", quote=FALSE)
head(n.res)
```

# Wrapping up

We report the session information.

```{r}

sessionInfo()
```
