---
title: "Single-cell RNA-seq of naive/primed embryonic stem cells: data preprocessing"
author: Tobias Messmer and Aaron Lun
date: 13 November 2017
output: 
  html_document:
    toc: true
    toc_float: true
    depth: 3
    number_sections: true
    theme: united 
    highlight: tango 
    fig_caption: false
---

```{r, echo=FALSE, results="hide"}
dir.create("figure-preprocess/", showWarning=FALSE)
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE, fig.path="figure-preprocess/")
options(bitmapType="cairo", width=100)
```

# Examining mapping and counting statistics

We have a look at the mapping statistics for the libraries in each batch.
This includes the proportion of reads that were mapped, and the proportion that were counted into genes.
The former should generally be at least 70-80% while the latter should be at least 50-60%.

```{r statcheck}
for (sample in c("2383", "2384", "2677", "2678", "2739", "2740")){  
    cur.file <- list.files("counts", pattern=paste0("my_qual_", sample), full=TRUE)
    qual.data <- read.table(cur.file, sep="\t", header=TRUE, row.names=1)
    total <- rowSums(qual.data)
    reads.mapped <- total - qual.data$Unassigned_Unmapped - qual.data$Unassigned_MappingQuality
    reads.assigned <- qual.data$Assigned
    cat(sprintf("Mapping percentages for batch %s:\n", sample))
    print(summary(reads.mapped/total*100))
    cat(sprintf("Counting percentages for batch %s:\n", sample))
    print(summary(reads.assigned/total*100))
    cat("\n")
}
```

# Creating a `SingleCellExperiment` object

## Reading in the counts 

The first step is to read in the counts and merge the technical replicates.
We start by merging the technical replicates within each batch (2383/2384).

```{r, datain}
library(edgeR)
all.counts <- list()
gene.names <- NULL
gene.length <- NULL
for (sample in c("2383", "2384", "2677", "2678", "2739", "2740")){
    cur.file <- list.files("counts", pattern=paste0("genic_counts_", sample), full=TRUE)
    current_counts <- read.table(cur.file, sep="\t", header=TRUE, row.names=1)

    # Checking gene names and length are the same as those in other files.
    if (is.null(gene.names)){
        gene.names <- rownames(current_counts)
        gene.length <- current_counts$Length
    } else {
        stopifnot(identical(gene.names, rownames(current_counts))) 
        stopifnot(identical(gene.length, current_counts$Length))
    }
    current_counts$Length <- NULL 
  
    # Take the technical replicates and merge them, if they exist.
    cellname <- colnames(current_counts)
    cellname <- sub("^lane[0-9]_", "", cellname)
    cellname <- sub("_L00[0-9]_", "_", cellname)
    cellname <- sub("_[12]$", "", cellname) 
    colnames(current_counts) <- cellname
    if (any(duplicated(colnames(current_counts)))) {
        current_counts <- sumTechReps(current_counts)
        gc()
    }

    # Adding to the list.
    all.counts[[sample]] <- as.matrix(current_counts)
}  
sapply(all.counts, ncol)
```

We then merge the replicates across batches (2677 + 2678, and 2739 + 2740).

```{r mergetech}
stopifnot(identical(colnames(all.counts[["2677"]]), colnames(all.counts[["2678"]])))
all.counts[["2677"]] <- all.counts[["2677"]] + all.counts[["2678"]]
all.counts[["2678"]] <- NULL
stopifnot(identical(colnames(all.counts[["2739"]]), colnames(all.counts[["2740"]])))
all.counts[["2739"]] <- all.counts[["2739"]] + all.counts[["2740"]]
all.counts[["2740"]] <- NULL
sapply(all.counts, ncol)
```

Finally, we `cbind`  everything together into one large matrix. 
This automatically adds the batch labels in front of the cell names.

```{r cbindmat}
for (sample in names(all.counts)) {
    current <- all.counts[[sample]] 
    colnames(current) <- paste0(sample, ".", colnames(current))
    all.counts[[sample]] <- current
}
combined.counts <- do.call(cbind, all.counts)
dim(combined.counts)
```

```{r, echo=FALSE, results='hide'}
rm(all.counts)
gc()
```

## Adding feature-level annotation 

Next, we add information such as gene symbols and chromosome locations.
We download annotation from BioMart and store it along with the length information.
For some reason, _biomaRt_ has decided to pad the Entrez ID; we get rid of that as well.

```{r, echo=FALSE, results="asis"}
library(BiocFileCache)
bfc <- BiocFileCache("biomart")
out <- bfcquery(bfc, "hg38.ensGene")
already.present <- nrow(out) > 0
```

```{r biomarting, eval=!already.present}
library(biomaRt)
ensembl <- useMart(biomart='ENSEMBL_MART_ENSEMBL', dataset="hsapiens_gene_ensembl", 
                   host="aug2017.archive.ensembl.org") # Ensembl version 90.
ensemblGenes <- getBM(attributes=c('ensembl_gene_id',  'chromosome_name', 'gene_biotype', 
                                   'external_gene_name', 'entrezgene'), filters="ensembl_gene_id", 
                                   values=gene.names, mart=ensembl) 
```

```{r, echo=FALSE}
if (already.present) {
    ensemblGenes <- readRDS(bfc[[out$rid[1]]])
} else {
    rpath <- bfcnew(bfc, "hg38.ensGene", ext="rds")
    saveRDS(file=rpath, ensemblGenes)
}
```


```{r featureData}
features <- ensemblGenes[match(gene.names, ensemblGenes$ensembl_gene_id),]
features$ensembl_gene_id <- gene.names
features$entrezgene <- as.character(gsub(" ", "", features$entrezgene))
row.names(features) <- gene.names
features$Length <- gene.length
head(features)
```

Mitochondrial genes and spike-in transcripts are defined.

```{r defineControls}
is.mito <- !is.na(features$chromosome_name) & features$chromosome_name=="MT"
summary(is.mito)
is.spike <- grepl("^ERCC", gene.names)
summary(is.spike)
```

## Adding sample-level annotation

We then add information about the phenotype of the cells. 
The cells are either in a native or primed state. 
Furthermore, cells are categorized by their respective batch number.

```{r phenData}
sample <- sub("^([0-9]+)\\..*", "\\1", colnames(combined.counts))
phenotype <- character(ncol(combined.counts))
phenotype[sample %in% c("2383", "2677", "2678")] <- "naive"
phenotype[sample %in% c("2384", "2739", "2740")] <- "primed"
table(phenotype)
batch <- character(ncol(combined.counts))
batch[sample %in% c("2383", "2384")] <- "1"
batch[sample %in% c("2677", "2678", "2739", "2740")] <- "2"
table(batch)
```

This is used to construct a phenotype table describing each sample.

```{r}
pheno <- data.frame(phenotype, batch, sample, row.names = colnames(combined.counts))
head(pheno)
```

## Constructing the `SingleCellExperiment` object

We create a object to facilitate further processing of the data.
We need to specify the spike-in set at this point.

```{r newSCESet}
library(SingleCellExperiment)
sce <- SingleCellExperiment(list(counts=as.matrix(combined.counts)), 
    rowData=as.matrix(features), colData=as.matrix(pheno))
isSpike(sce, "ERCC") <- is.spike 
sce
```

We also add a directory to store any useful results.

```{r}
resdir <- "results-preprocess"
dir.create(resdir, showWarning=FALSE)
```

```{r, echo=FALSE, results='hide'}
rm(combined.counts)
gc()
```

To enable a user-friendly analysis, it is convenient to exchange the ENSEMBL IDs with the name of the genes. 
For duplicated gene names, we merge them with the ENSEMBL IDs to avoid confusion.

```{r nameForID}
new.row.names <- as.vector(rowData(sce)$external_gene_name)
missing.name <- is.na(new.row.names)
new.row.names[missing.name] <- as.vector(rowData(sce)$ensembl_gene_id[missing.name])
dup.name <- duplicated(new.row.names)
new.row.names[dup.name] <-
  as.vector(paste0(rowData(sce)$external_gene_name[dup.name], "_",
         rowData(sce)$ensembl_gene_id[dup.name]))
rownames(sce) <- new.row.names
```

# Quality Control

## Removing low-quality cells

A number of QC metrics are calculated for each cell using the `calculateQCMetrics` function.
We also specify the set of feature controls that correspond to spike-in transcripts.

```{r QCMetrics}
library(scater)
sce <- calculateQCMetrics(sce, feature_controls=list(ERCC=is.spike, Mt=is.mito))
colnames(colData(sce))
```

Assuming that most cells are high-quality, we remove cells with outlier values for technical metrics.
We start with the library sizes and the number of expressed genes.
Cells with small library sizes (below ~100,000) or few expressed genes (below ~5000) are considered to be of low quality.
We do the same for the proportion of reads mapped to mitochondrial and spike-in transcripts.
Cells with high values for either metric are likely to be poor-quality, as cytoplasmic/nuclear RNA has been lost.


```{r histoSize, fig.width=10, fig.height=10}
multiplot(cols=2,
    plotPhenoData(sce, aesth=aes_string(x="sample", y="log10_total_counts")),
    plotPhenoData(sce, aesth=aes_string(x="sample", y="total_features")),
    plotPhenoData(sce, aesth=aes_string(x="sample", y="pct_counts_ERCC")),
    plotPhenoData(sce, aesth=aes_string(x="sample", y="pct_counts_Mt"))
)
```

To filter out low-quality cells, we identify those with outlier values for any of the above metrics.
We do this for each batch separately due to differences in sequencing depth and spike-in proportion between batches.

```{r, remOutliers1}
libsize.drop <- isOutlier(sce$total_counts, nmads=3, type="lower", log=TRUE, batch=sce$sample) 
feature.drop <- isOutlier(sce$total_features, nmads=3, type="lower", log=TRUE, batch=sce$sample) 
mito.drop <- isOutlier(sce$pct_counts_Mt, nmads=3, type="higher", batch=sce$sample)
spike.drop <- isOutlier(sce$pct_counts_ERCC, nmads=3, type="higher", batch=sce$sample)
discard <- libsize.drop | feature.drop | spike.drop | mito.drop
data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop), ByMito=sum(mito.drop),
           BySpike=sum(spike.drop), Remaining.in.batch=sum(!discard))
```

We remove these cells from our `sce` object prior to further analysis.

```{r, remOutliers3}
sce <- sce[, !discard]
dim(sce) 
```

```{r, echo=FALSE, results='hide'}
gc()
```

## Classification into cell cycle phase

As part of the quality checks, we can also examine the distribution of cells into different cell cycle phases.

```{r filterPhase}
library(scran)
hs.pairs <- readRDS(system.file("exdata", "human_cycle_markers.rds", package="scran"))
assigned <- cyclone(sce, pairs=hs.pairs, gene.names=rowData(sce)$ensembl_gene_id)
write.table(file=file.path(resdir, "phases.tsv"), assigned, row.names=FALSE, sep="\t", quote=TRUE)
head(assigned$scores)
```

Visualizing the scores.

```{r phaseplot, fig.width=10, fig.height=6}
par(mfrow=c(1,2))
plot(assigned$scores$G1, assigned$scores$G2M, xlab="G1 score", ylab="G2/M score", pch=16)
plot(assigned$scores$G1, assigned$scores$S, xlab="G1 score", ylab="S score", pch=16)
```

Phase assignments are stored in the `sce` object prior to further analysis.

```{r}
sce$phase <- assigned$phases
table(assigned$phases)
```

# Examining the genes

We have a look at the average abundance of the genes.

```{r aveGenes}
ave.counts <- calcAverage(sce)
rowData(sce)$AveCount <- ave.counts
hist(log10(ave.counts), breaks=100, main="", col="grey", xlab=expression(Log[10]~"average count"))
```

We have a look at the number of cells expressing each gene.

```{r numGenes}
ncells <- nexprs(sce, byrow=TRUE)
rowData(sce)$Ncells <- ncells
plot(ave.counts, ncells, xlab="Average count", ylab="Number of cells", log="x", pch=16, cex=0.5)
```

Inspecting the 50 most highly expressed genes should show constitutively expressed genes like mitochondrial transcripts, ribosomal proteins, actin and other usual suspects.

```{r highEx, fig.height=12, fig.width=6}
fontsize <- theme(axis.text=element_text(size=12), axis.title=element_text(size=16))
plotQC(sce, type = "highest-expression", n=50) + fontsize
```

We discard genes that are not expressed at all.
Such genes provide no information for characterizing heterogeneity across the population.

```{r}
keep <- ave.counts > 0
sce <- sce[keep,]
summary(keep)
```

# Normalization of various technical biases

## Removing scaling biases

Cells are clustered together and scaling biases are removed using the deconvolution method.
This is done using only high-abundance genes to avoid problems with having too many zeroes in low-abundance genes.

```{r sumnorm}
keep <- rowData(sce)$AveCount >= 1
clusters <- quickCluster(sce, method="igraph", subset.row=keep)
sce <- computeSumFactors(sce, cluster=clusters, subset.row=keep, min.mean=NULL)
summary(sizeFactors(sce))
```

This correlates well with library size normalization, which is typically expected from read count data.

```{r sumnormplot}
plot(sizeFactors(sce), sce$total_counts/1e6, log="xy", ylab="Library size (millions)", 
    xlab="Size factor", col=ifelse(sce$phenotype=="naive", "black", "grey"))
```

Spike-in transcripts are normalized separately, because they aren't subject to the effects of total RNA content.
Here, the correlation is worse, which is also expected.

```{r spikenorm}
sce <- computeSpikeFactors(sce, type="ERCC", general.use=FALSE) 
plot(sizeFactors(sce), sizeFactors(sce, type="ERCC"), log="xy", ylab="Spike-in size factor", 
    xlab="Deconvolution size factor", col=ifelse(sce$phenotype=="naive", "black", "grey"))
```

Finally, the size factors are applied to compute normalized log-expression values.

```{r}
sce <- normalize(sce)
```

## Checking for systematic technical effects

As the cells have been sequenced in different batches, a batch-related bias is expected.
We can examine the proportion of variance attributable to various factors.
Phenotype is the largest contributor (obviously) but the batch effect also needs to be considered.

```{r varbias}
plotExplanatoryVariables(sce, variables=c("phenotype", "batch")) + fontsize
```

The spike-in variance measures any leftover technical bias from scaling normalization, assuming that library quantification was _not_ performed.
In this case, the spike-ins are correlated with batch, so let's get rid of that first.
After doing so, you can see that there's not a lot of leftover bias.

```{r}
library(limma)
alt.sce <- sce
logcounts(alt.sce) <- removeBatchEffect(logcounts(alt.sce), alt.sce$batch)
plotExplanatoryVariables(alt.sce, variables=c("total_counts_ERCC",
    "log10_total_counts_ERCC")) + fontsize
```

# Wrapping up

We save the object to file for use in downstream analyses.

```{r}
saveRDS(sce, file="sce_all.rds")
sessionInfo()
```

