---
title: "Single-cell RNA-seq Analysis of naive/primed embryonic stem cells: remapping naivete"
author: Tobias Messmer and Aaron Lun
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    depth: 3
    number_sections: true
    theme: united 
    highlight: tango 
    fig_caption: false
---

```{r, echo=FALSE, results="hide"}
dir.create("figure-bulk/", showWarning=FALSE)
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE, fig.path="figure-bulk/")
options(bitmapType="cairo", width=100)
```

# Locating ESCs according to their naive/primed-ness

Finally, we want to validate the markers that we identified in ... in other cell lines.
For this, we used published bulkRNA seq data sets of naive and primed hESCs.

- Human ESCs:
    - __Publication:__ Petropoulos et al. (2016), _Cell_. Single-cell RNA-seq reveals lineage and X chromosome dynamics in human preimplantation embryos. 
    - __Data source:__ `E-MTAB-3929.processed.1.zip` at http://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-3929
- Mouse ESCs:
    - __Publication:__ Mohammed et al. (2017), _Cell Reports_. Single-Cell Landscape of Transcriptional Heterogeneity and Cell Fate Decisions during Mouse Early Gastrulation
    - __Data source:__ `GSE100597_count_table_QC_filtered.txt.gz` at https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE100597


The aim is to identify differentially expressed genes in those data sets and see if we have them as well.

```{r}
resdir <- "results-bulk"
dir.create(resdir, showWarning=FALSE)
```

# Data input

First, read in the data.

```{r bulk.in}
pastor16 <- read.table(file.path("../data/other_counts", "bulkRNA", "Pastor2016_rawReadcountRNAseq.txt"), sep = "\t", header = TRUE)
gene.names <- pastor16$Probe
pastor16 <- as.matrix(pastor16[,14:19])
rownames(pastor16) <- gene.names
keep <- rowSums(is.na(pastor16))==0
pastor16 <- pastor16[keep,]
dim(pastor16)
```

Next, we test for differences in expression between the naive and primed conditions.
This is done using _edgeR_, after summing all cells within each batch to create a pseudo-bulk sample.

```{r desetup}
library(edgeR)
y <- DGEList(pastor16)
ptype <- factor(c('Primed', 'Primed', 'Primed', 'Naive', 'Naive', 'Naive'))
design_pheno <- model.matrix(~0 + ptype) 
keep <- aveLogCPM(y) > aveLogCPM(5, mean(y$samples$lib.size))
y <- y[keep,]
y <- calcNormFactors(y)
y <- estimateDisp(y, design_pheno)
fit <- glmQLFit(y, design_pheno, robust=TRUE)
con <- makeContrasts(ptypeNaive - ptypePrimed, levels = design_pheno)
res <- glmQLFTest(fit, contrast=con)
DE_pastor <- topTags(res, n=Inf)$table
#DE_pastor <- DEmarkers[DEmarkers$FDR <= 0.05,]
```

We now repeat the same pipeline for the other data sets as well.

```{r}
theunissen16 <- read.table(file.path("../data/other_counts", "bulkRNA", "Theunissen2016_rawReadcountRNAseq.txt"), sep = "\t", header = TRUE)
gene.names <- theunissen16$Probe
theunissen16 <- as.matrix(theunissen16[,14:23])
rownames(theunissen16) <- gene.names
keep <- rowSums(is.na(theunissen16))==0
theunissen16 <- theunissen16[keep,]
dim(theunissen16)
```


```{r}
y <- DGEList(theunissen16)
ptype <- factor(c('Primed', 'Naive', 'Naive', 'Naive', 'Naive', 'Naive', 'Naive', 'Primed', 'Primed', 'Primed'))
design_pheno <- model.matrix(~0 + ptype) 
keep <- aveLogCPM(y) > aveLogCPM(5, mean(y$samples$lib.size))
y <- y[keep,]
y <- calcNormFactors(y)
y <- estimateDisp(y, design_pheno)
fit <- glmQLFit(y, design_pheno, robust=TRUE)
con <- makeContrasts(ptypeNaive - ptypePrimed, levels = design_pheno)
res <- glmQLFTest(fit, contrast=con)
summary(decideTestsDGE(res))
DE_theunissen <- topTags(res, n=Inf)$table
#DE_theunissen <- DEmarkers[DEmarkers$FDR <= 0.05,]
```

Now that we have the DE genes of the bulk RNA sets, we can load in our identified markers and plot them.

```{r}
naive.markers <- readRDS(file=file.path('results-remapping', "naive_markers.rds"))
primed.markers <- readRDS(file=file.path('results-remapping', "primed_markers.rds"))

naive.markers <- naive.markers[naive.markers %in% rownames(DE_pastor) & naive.markers %in% rownames(DE_theunissen)]
primed.markers <- primed.markers[primed.markers %in% rownames(DE_pastor) & primed.markers %in% rownames(DE_theunissen)]

bound <- 10

naive_mat <- data.frame(Pastor = DE_pastor[naive.markers,]$logFC, Theunissen = DE_theunissen[naive.markers,]$logFC)
rownames(naive_mat) <- naive.markers
naive_mat[which(naive_mat>=bound, arr.ind = TRUE)] <- bound
naive_mat[which(naive_mat< (-bound), arr.ind = TRUE)] <- -bound

primed_mat <- data.frame(Pastor = DE_pastor[primed.markers,]$logFC, Theunissen = DE_theunissen[primed.markers,]$logFC)
rownames(primed_mat) <- primed.markers 
primed_mat[which(primed_mat>=bound, arr.ind = TRUE)] <- bound
primed_mat[which(primed_mat< (-bound), arr.ind = TRUE)] <- -bound

library(pheatmap)
anno <- data.frame(Marker = c(rep('naive', length(naive.markers)), rep('primed', length(primed.markers))))
rownames(anno) <- c(naive.markers, primed.markers)
pheatmap(rbind(naive_mat, primed_mat), annotation_row = anno, cluster_cols = FALSE)

```
