---
title: "Single-cell RNA-seq of naive/primed embryonic stem cells: differential variability" 
author: Tobias Messmer and Aaron Lun
date: 13 November 2017
output: 
  BiocStyle::html_document:
    toc: true
    toc_float: true
    depth: 3
    number_sections: true
    theme: united 
    highlight: tango 
    fig_caption: false
---

```{r, echo=FALSE, results="hide"}
dir.create("figure-diff/", showWarning=FALSE)
knitr::opts_chunk$set(error=FALSE, warning=FALSE, message=FALSE, fig.path="figure-diff/")
```

# Data preparation

The idea here is to test for whether there are any changes in gene expression of germ layer-specific markers within the naive and the primed population, respectively.
This will allow us to determine whether the primed cells prepare for differentiation into the three germ layers.
We focus on the genes in the germ layer to maximize our power to seee any heterogeneity that might be present.
The logic is that, if the populations are really homogeneous, even a highly biased view should not yield any structure.

We load in the lineage markers that we'll use for focused examination of heterogeneity.

```{r get.list}
by.lineage <- read.csv(file.path("../data/marker", "commit_lineages.csv"), 
    header=TRUE, stringsAsFactors=FALSE, sep= ';')
head(by.lineage)
```

```{r adj.list}
library(org.Hs.eg.db)
ensembls <- mapIds(org.Hs.eg.db, keytype="SYMBOL", 
    keys=by.lineage$Symbol, column="ENSEMBL")

library(SingleCellExperiment)
sce_naive <- readRDS("results-naive/sce_naive.rds")
chosen <- ensembls %in% rowData(sce_naive)$ensembl_gene_id 
table(chosen, by.lineage$Germ_layer)

by.lineage <- by.lineage[chosen,]
keep <- rowData(sce_naive)$ensembl_gene_id %in% ensembls
```

We also set up an output directory.

```{r def.dir}
resdir <- "results-diff"
dir.create(resdir, showWarning=FALSE)
```

# Heterogeneity in the naive population 

We remove the transition cells from the naive subpopulation to ensure that they don't contribute to heterogeneity.

```{r}
groups <- read.table("results-transition/groups.tsv", 
    header=TRUE, stringsAsFactors=FALSE)
is.transition <- groups$Cell[groups$Type=="transition"]
sce_naive <- sce_naive[,!colnames(sce_naive) %in% is.transition]
ncol(sce_naive)
```

We calculate the PCs from the subset of lineage markers.
It is important not to forget to remove the batch effect to ensure that they don't distord the clustering. 

```{r naive.vartrend}
library(scater)
sce_naive <- normalize(sce_naive)

library(limma)    
norm_exprs(sce_naive) <- removeBatchEffect(exprs(sce_naive), batch=sce_naive$batch) 
sce_naive <- runPCA(sce_naive, subset_row=keep, exprs_values="norm_exprs", scale=FALSE)
```

We now want to further assess whether clusters form in the naive population if we only use genes that are specific for the three germlayers. 
We use PCA and tSNE to visualize potential clustering, and we can colour the points by expression of specific markers of mesoderm, endoderm and ectoderm.
This shows whether the distribution of cells in each embedding is related to the lineage.

```{r pca.naive}
library(ggplot2)
fontsize <- theme(axis.text=element_text(size=12), 
    axis.title=element_text(size=16), title = element_text(size=12))

pca_naive1 <- plotPCA(sce_naive, colour_by = 'SNAI1') + fontsize # mesoderm
pca_naive2 <- plotPCA(sce_naive, colour_by = 'ITGA6') + fontsize # ectoderm
pca_naive3 <- plotPCA(sce_naive, colour_by = 'PAF1') + fontsize # endoderm
pca_naive4 <- plotPCA(sce_naive, colour_by = 'CDK1') + fontsize # cell cycle

multiplot(pca_naive1, pca_naive2, pca_naive3, pca_naive4, cols = 2)
```

There are no distinct clusters visible, so we assume that cells are still in an uncommited stage. 
We want to see whether the same results can be obtained via t-SNE.

```{r tsne.naive}
set.seed(100)
sce_naive <- runTSNE(sce_naive, subset_row=keep, scale=FALSE, perplexity=30)

#mesoderm
tsne_naive1 <- plotTSNE(sce_naive, colour_by= "SNAI1") + fontsize
saveRDS(tsne_naive1, file = file.path(resdir, 'naive_tsne1.Rds'))

#ectoderm
tsne_naive2 <- plotTSNE(sce_naive, colour_by= "ITGA6") + fontsize
saveRDS(tsne_naive2, file = file.path(resdir, 'naive_tsne2.Rds'))

#endoderm
tsne_naive3 <- plotTSNE(sce_naive, colour_by= "PAF1") + fontsize
saveRDS(tsne_naive3, file = file.path(resdir, 'naive_tsne3.Rds'))

#cell cycle
tsne_naive4 <- plotTSNE(sce_naive, colour_by= 'CDK1') + fontsize
saveRDS(tsne_naive4, file = file.path(resdir, 'naive_tsne4.Rds'))

multiplot(tsne_naive1, tsne_naive2, tsne_naive3, tsne_naive4, cols = 2)
```

# Heterogeneity in the primed subpopulation

We repeat the same analysis for the primed population.

```{r primed.vartrend, fig.height=6, fig.width=10}
sce_primed <- readRDS("results-primed/sce_primed.rds")
sce_primed <- normalize(sce_primed)
norm_exprs(sce_primed) <- removeBatchEffect(exprs(sce_primed), batch=sce_primed$batch) 
sce_primed <- runPCA(sce_primed, subset_row=keep, exprs_values="norm_exprs", scale=FALSE)
```

We perform dimensionality reduction only on lineage markers and colour based on their expression. 
First, PCA.

```{r pca.primed}
pca_primed1 <- plotPCA(sce_primed, colour_by = 'SNAI1') + fontsize # mesoderm
pca_primed2 <- plotPCA(sce_primed, colour_by = 'ITGA6') + fontsize # ectoderm
pca_primed3 <- plotPCA(sce_primed, colour_by = 'PAF1') + fontsize # endoderm
pca_primed4 <- plotPCA(sce_primed, colour_by = 'CDK1') + fontsize # cell cycle
multiplot(pca_primed1, pca_primed2, pca_primed3, pca_primed4, cols = 2)
```

Then, we do the same with t-SNE.

```{r tsne.primed}
set.seed(100)
sce_primed <- runTSNE(sce_primed, perplexity=30, subset_row=keep, scale=FALSE)

#mesoderm
tsne_primed1 <- plotTSNE(sce_primed, colour_by= "SNAI1") + fontsize
saveRDS(tsne_primed1, file = file.path(resdir, 'primed_tsne1.Rds'))

#ectoderm
tsne_primed2 <- plotTSNE(sce_primed, colour_by= "ITGA6") + fontsize
saveRDS(tsne_primed2, file = file.path(resdir, 'primed_tsne2.Rds'))

#endoderm
tsne_primed3 <- plotTSNE(sce_primed, colour_by= "PAF1") + fontsize
saveRDS(tsne_primed3, file = file.path(resdir, 'primed_tsne3.Rds'))

#cell cycle
tsne_primed4 <- plotTSNE(sce_primed, colour_by= 'CDK1') + fontsize
saveRDS(tsne_primed4, file = file.path(resdir, 'primed_tsne4.Rds'))

multiplot(tsne_primed1, tsne_primed2, tsne_primed3, tsne_primed4, cols = 2)
```

# Testing for correlations

To be more quantitative, we compare the pairwise correlations within each condition.
First, naive:

```{r}
library(scran)
cor.out_naive <- correlatePairs(sce_naive, subset.row=keep)
sig.cor_naive <- cor.out_naive$FDR <= 0.05
summary(sig.cor_naive)
```

We can examine between which layers these correlations are occurring:

```{r}
LAYER <- function(vals) {
    by.lineage$Germ_layer[match(vals, by.lineage$Symbol)]
}

sig.cor_naive_pos <- cor.out_naive[sig.cor_naive & cor.out_naive$rho > 0,]
table(LAYER(sig.cor_naive_pos$gene1), LAYER(sig.cor_naive_pos$gene2))

sig.cor_naive_neg <- cor.out_naive[sig.cor_naive & cor.out_naive$rho < 0,]
table(LAYER(sig.cor_naive_neg$gene1), LAYER(sig.cor_naive_neg$gene2))
```

We repeat this for the primed population:

```{r}
cor.out_primed <- correlatePairs(sce_primed, subset.row=keep)
sig.cor_primed <- cor.out_primed$FDR <= 0.05
summary(sig.cor_primed)

sig.cor_primed_pos <- cor.out_primed[sig.cor_primed & cor.out_primed$rho > 0,]
table(LAYER(sig.cor_primed_pos$gene1), LAYER(sig.cor_primed_pos$gene2))

sig.cor_primed_neg <- cor.out_primed[sig.cor_primed & cor.out_primed$rho < 0,]
table(LAYER(sig.cor_primed_neg$gene1), LAYER(sig.cor_primed_neg$gene2))
```

There is a minor increase in the number of correlated gene pairs in the primed condition.
However, there is no evidence for an enrichment in positive intra-layer correlations and negative inter-layer correlations,
which would be expected if commitment into the germ layers was beginning to occur.

# Wrapping up

Returning the session information:

```{r end}
sessionInfo()
```
